sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c

```diff
@@ -248,6 +248,129 @@ static int _gnrc_tcp_open(gnrc_tcp_tcb_t *tcb, char *target_addr, uint16_t targe
+                     uint16_t port, uint16_t netif)
+{
+#ifdef MODULE_GNRC_IPV6
+    if (family != AF_INET6) {
+        return -EAFNOSUPPORT;
+    }
+
+    if (addr == NULL && addr_size == 0) {
+        ipv6_addr_set_unspecified((ipv6_addr_t *) ep->addr.ipv6);
+    }
+    else if (addr_size == sizeof(ipv6_addr_t)) {
+        memcpy(ep->addr.ipv6, addr, sizeof(ipv6_addr_t));
+    }
+    else {
+        return -EINVAL;
+    }
+#else
+    return -EAFNOSUPPORT;
+#endif
+
+    ep->family = family;
+    ep->port = port;
+    ep->netif = netif;
+    return 0;
+}
+
+int gnrc_tcp_ep_from_str(gnrc_tcp_ep_t *ep, const char *str)
+{
+    assert(str);
+
+    unsigned port = 0;
+    unsigned netif = 0;
+
+    /* Examine given string */
+    char *addr_begin = strchr(str, '[');
+    char *addr_end = strchr(str, ']');
+
+    /* 1) Ensure that str contains a single pair of brackets */
+    if (!addr_begin || !addr_end || strchr(addr_begin + 1, '[') || strchr(addr_end + 1, ']')) {
+        return -EINVAL;
+    }
+    /* 2) Ensure that the first character is the opening bracket */
+    else if (addr_begin != str) {
+        return -EINVAL;
+    }
+
+    /* 3) Examine optional port number */
+    char *port_begin = strchr(addr_end, ':');
+    if (port_begin) {
+        /* 3.1) Ensure that there are characters left to parse after ':'. */
+        if (*(++port_begin) == '\0') {
+            return -EINVAL;
+        }
+
+        /* 3.2) Ensure that port is a number (atol, does not report errors) */
+        for (char *ptr = port_begin; *ptr; ++ptr) {
+            if ((*ptr < '0') || ('9' < *ptr)) {
+                return -EINVAL;
+            }
+        }
+
+        /* 3.3) Read and verify that given number port is within range */
+        port = atol(port_begin);
+        if (port > 0xFFFF) {
+            return -EINVAL;
+        }
+    }
+
+    /* 4) Examine optional interface identifier. */
+    char *if_begin = strchr(str, '%');
+    if (if_begin) {
+        /* 4.1) Ensure that the identifier is not empty and within brackets. */
+        if (addr_end <= (++if_begin)) {
+            return -EINVAL;
+        }
+
+        /* 4.2) Ensure that the identifier is a number (atol, does not report errors) */
+        for (char *ptr = if_begin; ptr != addr_end; ++ptr) {
+            if ((*ptr < '0') || ('9' < *ptr)) {
+                return -EINVAL;
+            }
+        }
+
+        /* 4.3) Read and replace addr_end with if_begin. */
+        netif = atol(if_begin);
+        addr_end = if_begin - 1;
+    }
+
+#ifdef MODULE_GNRC_IPV6
+    /* 5) Try to parse IP Address. Construct Endpoint on after success. */
+    char tmp[IPV6_ADDR_MAX_STR_LEN];
+
+    /* 5.1) Verify address length and copy address into temporary buffer.
+     *      This is required to preserve constness of input.
+     */
+    int len = addr_end - (++addr_begin);
+
+    if (0 <= len && len <= (int) IPV6_ADDR_MAX_STR_LEN) {
+        memset(tmp, 0, sizeof(tmp));
+        memcpy(tmp, addr_begin, len);
```

Comment:
If you specify the function as destructive to `str` (is there a reason to keep it?) you could ditch the copy operation as a whole and just set

```C
*addr_end = `\0`
```

Reply:
The reason is simply that I do allow users to supply a string constant.
This function should be usable like this:
```
gnrc_tcp_ep_t ep;
gnrc_tcp_ep_from_str(&ep, "[::]");
```

The above code snipped would compile but will lead to a runtime error since string constants are usually placed in the read only section of a program. Keeping the constness of str prevents runtime errors. That the reasoning behind the signature.

If you want to test it:

```
int main(int argc, char **argv) {
    char *foo = "Foo";
    foo[1] = '\0';
    return 0;
}
```

Compiles but causes a segfault on execution.
`